### ISSUE TITLE
Share feature proposal

### ISSUE THREAD

**Issue by smoya** [asyncapi/studio#127]
#### Reason/Context
The idea of this proposal is to add a "Share" button (or any component that fits) allowing the user to generate a URL that can be shared with anyone, and that will open the Studio but with the desired spec file loaded, representing that file in a certain moment in time, and loading any Studio setting the person who shared configured by then as well.

Sharing a URL that points to a reproducible Studio state is useful for several scenarios:

- Users can share AsycnAPI files without having to host them somewhere (like GH gists, etc), plus the fact you have all Studio amazing features out of the box.
- Tools that generate AsyncAPI files (like [Chatbot](https://github.com/asyncapi/chatbot), CLI...) could give a URL as a result instead of generating the code, or printing to stdout, etc.

You can see an example of such feature in Go (language) playground: https://play.golang.org/p/ENxYVMoY9VZ

#### Description
Let's see first a couple of mocks to illustrate the idea:
![Mock share button](https://user-images.githubusercontent.com/1083296/138260981-276f886e-9182-41e9-9ae0-c8d5271de46b.png)

![Mock modal](https://user-images.githubusercontent.com/1083296/138262341-709b7d77-cf52-47d0-8a4c-a28d852bd7b1.png)

This is not as trivial as it might seem. I'm gonna add some context:

The [first idea](https://github.com/asyncapi/playground/issues/202) was to just accept all state (spec file, settings...) encoded as base64 in a query param. Then the studio will decode that value to obtain the state and draw it. 
The problem with this approach, as @magicmatatjahu [stated](https://github.com/asyncapi/playground/issues/202#issuecomment-947507773), is that each browser have their own limitations on the URL length (also with the length of each query parameter). Most of them, limiting the full URL to 2048 chars which is not a lot. E.g. [Streetlights MQTT example](https://github.com/asyncapi/spec/blob/master/examples/streetlights-mqtt.yml) has ~6000 chars already.

With such limitation in place, there are not a lot of solutions for solving this. 

#### Proposal
**TL;DR: Storing the state in an external storage. Downloading it later and import it into the Studio.** 

> [!IMPORTANT]  
> An alternative implementation, "cheapest" and simple-enough is mentioned in https://github.com/asyncapi/studio/issues/127#issuecomment-2074404179

![Simple diagram](https://user-images.githubusercontent.com/1083296/138279043-2f02efe3-2f96-4137-b910-86688fb8229e.png)

- Generating the URL will require passing the current state (AsyncAPI file + any settings) to an external service, which will generate a **UUID** (either based on the content like SHA256, or a UUID v(x). To be defined) and store the given state in a Storage (external database, cloud storage, etc). That **UUID** will be returned to the Studio and will be the only identifier needed to construct the shareable URL. Optionally, to use (or maybe create our own?) a URL shortener service for making the URL shorter.
- Loading a shared URL will require a call to an external service (same), which from a given **UUID**, will fetch the state from a Storage (external database, cloud storage, etc) and return it to the Studio, which will load such state.

Note: Other services like [Chatbot](https://github.com/asyncapi/chatbot/issues/26) could also use this service.

Obviously, this is only the big picture. Details should be clarify. Further investigation should be done, especially around the following topics among others:

- What storage should we use? How can we get budget for it?
- Shall we use https://github.com/asyncapi/server-api as server for making the sharing logic? Or rather a separate microservice that just owns such responsibility?
- Security concerns. Only available to our services.
- Monitoring seems now like a must, as it is a server giving a functionality to a public hosted tool. High Availability is a must as well.
- Which protocol to use? GRPC?
- Etc.

cc @magicmatatjahu @derberg

**Comment by aayushmau5** [comment]
Love this idea a lot. It would be interesting to see how we tackle these problems. 

> What storage should we use? How can we get budget for it?
Shall we use https://github.com/asyncapi/server-api as server for making the sharing logic? Or rather a separate microservice that just owns such responsibility?

I have a feeling that we can go serverless for this approach(like cloudflare workers and cloudflare KV as a storage). Though we have to see the costing a bit, but I think it's really easy to get up and running with this approach.

**Comment by boyney123** [comment]
Yeah second what @aayushmau5 says, I like this idea @smoya. Sharing your state (doc) with others and other tools becomes very interesting.

> I have a feeling that we can go serverless for this approach

Yeah guess loads of ideas we could go, but if you wanted you could go with a simple AWS Serverless solution that uses API Gateway + DyanmoDB (or s3) for example to store the data (even put a TTL on the data in the table so it expires/deletes after X days). Also, they offer quite a big free tier there. (Any hosting would work just need to figure out where as you said)

**Comment by AceTheCreator** [comment]
@smoya how's it going with this? 

Any help needed? 

I resumed development on the chatbot and I could really use this feat.

**Comment by smoya** [comment]
@AceTheCreator AFAIK nobody is working on this feature yet.

**Comment by AceTheCreator** [comment]
@smoya no worries. I'll look into it and come up with something, then we can iterate from there. what do you think?

**Comment by smoya** [comment]
> @smoya no worries. I'll look into it and come up with something, then we can iterate from there. what do you think?

Do you feel it falls into https://www.asyncapi.com/roadmap? If it does, I think you are free to pick it up (well we are always free to pick up whatever, but always convenient to follow the roadmap, right?).

**Comment by smoya** [comment]
still relevant

**Comment by smoya** [comment]
Since [Netlify Blobs](https://docs.netlify.com/blobs/overview/) are available to general public, I believe we could use them to store the AsyncAPI docs when clicking on the "share" button (by calling a Netlify serverless function which does the store).
Later, we could retrieve it inside a [Netlify Edge Function](https://docs.netlify.com/edge-functions/overview/), and somehow inject it in the final rendered page and use it (this should be tested, not sure if will work but seems so) as in the following example https://www.netlify.com/blog/add-personalization-to-static-html-with-edge-functions-no-browser-javascript/

**Comment by smoya** [comment]
> This seems a good solution, any limits on no. of blobs stored ?

I couldn't find such info. I left a question in Netlify forums just in case: https://answers.netlify.com/t/blobs-pricing-and-limits/119907

**Comment by smoya** [comment]
> > This seems a good solution, any limits on no. of blobs stored ?
> 
> I couldn't find such info. I left a question in Netlify forums just in case: https://answers.netlify.com/t/blobs-pricing-and-limits/119907

Got [an answer](https://answers.netlify.com/t/blobs-pricing-and-limits/119907/3?u=sergio). Not written in stone but enough for us I would say.

>  can’t officially promise you that this is final, but in case you’re cautious about using this before knowing how much it will cost, Blobs will be likely priced something like this: Customers get 100GB for free, and additional GBs will cost 9 cents a month. Things might change, but that’s the ballpark.

**Comment by ibishal** [comment]
Hey @smoya , can we start developing on this as we have decided to go with netlify blob as an external storage ?

**Comment by smoya** [comment]
> Hey @smoya , can we start developing on this as we have decided to go with netlify blob as an external storage ?

Hi @ibishal, the last decision is not on my plate but on Studio code owners. Anyway I guess you are more than free to work on your side and submit a PR.

**Comment by KhudaDad414** [comment]
Hi @ibishal please feel free to open a PR.
